### PR TITLE
1247 document sync

### PR DESCRIPTION
**Pull request by clemens-msupply**
Part of #1247 

# 👩🏻‍💻 What does this PR do? 
- sync + test `form_schema` table (was missing so I quickly added it)
- sync `document_registry` table
- sync `document` table (this is not tested yet and will be finished in the next PR)
- some missed renames for `timestamp`  -> `datetime`

# Testing
- Just run the integration tests for `form_schema` and `document_registry` (however integration tests don't compile in this branch but fixes should be done upstream... so maybe just do a code review and do a full test in next PR)

### CODE REVIEW

File: server/repository/migrations/postgres/2022-04-27T13-05_create_document_trigger/up.sql [hunk 1]
```diff
@@ -0,0 +1,5 @@
+ALTER TYPE changelog_table_name ADD VALUE 'document' AFTER 'clinician_store_join';
+
+CREATE TRIGGER document_trigger
```
Comment (roxy-dao) [inline server/repository/migrations/postgres/2022-04-27T13-05_create_document_trigger/up.sql:3]: Will need to create a function for document trigger if you want to use the is_sync_update flag

File: server/repository/migrations/sqlite/2022-04-27T13-05_create_document_trigger/up.sql [hunk 2]
```diff
@@ -0,0 +1,20 @@
+CREATE TRIGGER document_insert_trigger
+  AFTER INSERT ON document
+  BEGIN
+    INSERT INTO changelog (table_name, record_id, row_action)
```
Comment (roxy-dao) [inline server/repository/migrations/sqlite/2022-04-27T13-05_create_document_trigger/up.sql:4]: add is_sync_update?

Comment (clemens-msupply) [reply server/repository/migrations/sqlite/2022-04-27T13-05_create_document_trigger/up.sql:4]: thanks there is a second PR on top of this one coming and will add this at the end!

File: server/service/src/programs/patient/patient_updated.rs [hunk 3]
```diff
@@ -35,7 +61,7 @@ pub fn patient_document_updated(
         r#type: NameType::Patient,
         is_customer: true,
         is_supplier: false,
-        supplying_store_id: Some(store_id.to_string()),
+        supplying_store_id: None,
```
Comment (roxy-dao) [inline server/service/src/programs/patient/patient_updated.rs:64]: How would the patient's home store be saved if they're created in remote? 🤔

File: server/service/src/programs/patient/update.rs [hunk 4]
```diff
@@ -78,6 +73,15 @@ pub fn update_patient(
                     DocumentInsertError::InvalidParent(_) => UpdatePatientError::InvalidParentId,
                 })?;
 
+            if is_latest_doc(ctx, service_provider, &doc)
+                .map_err(UpdatePatientError::DatabaseError)?
+            {
+                // update the names table
+                patient_document_updated(&ctx.connection, &doc_timestamp, patient)?;
```
Comment (roxy-dao) [inline server/service/src/programs/patient/update.rs:80]: Nit: Rename function? The patient table is being updated, not the patient's document?

Comment (clemens-msupply) [reply server/service/src/programs/patient/update.rs:80]: I used it in a "callback" style, i.e. the caller just says: "the patient document has been updated; now do what every is needed". Its like sending an event. However, think I will rename it to `update_patient_name()` anyway...

File: server/service/src/programs/patient/patient_updated.rs (hunk 3 above)
Comment (clemens-msupply) [reply server/service/src/programs/patient/patient_updated.rs:64]: Good point, I create a new PR where I take over values from the existing name row!

File: server/service/src/document/mod.rs [hunk 5]
```diff
@@ -1,28 +1,20 @@
...
         .document_service
         .document(ctx, &doc.name, None)?;
-    if let Some(lastest_existing) = latest_existing {
-        if lastest_existing.timestamp > doc.timestamp {
-            // newer doc already exist
-            return Ok(false);
-        }
-    }
-    Ok(true)
+    Ok(latest_existing.map(|e| e.id == doc.id).unwrap_or(false))
```
Comment (andreievg) [inline server/service/src/document/mod.rs:19]: Just looking at the code and method description, can't quite see how latest check it's done, it looks like `check if document id is unique for document name`

File: server/service/src/programs/encounter/insert.rs [hunk 6]
```diff
@@ -58,32 +58,7 @@ pub fn insert_encounter(
...
-                    UpdateProgramDocumentError::DatabaseError(err) => {
-                        InsertEncounterError::DatabaseError(err)
-                    }
-                    UpdateProgramDocumentError::InternalError(err) => {
-                        InsertEncounterError::InternalError(err)
-                    }
-                })?;
-            }
-
-            // Updating the document will trigger an update in the patient (names) table
```
Comment (andreievg) [inline server/service/src/programs/encounter/insert.rs:85]: This comment was removed is that right ?

File: server/service/src/document/mod.rs (hunk 5 above)
Comment (clemens-msupply) [reply server/service/src/document/mod.rs:19]: Yes thanks, this is buggy, it has already been updated in a later PR where I compare the timestamp...

File: server/service/src/programs/encounter/update.rs [hunk 7]
```diff
@@ -51,16 +51,38 @@ pub fn update_encounter(
             let encounter_start_datetime = encounter.start_datetime;
             let doc = generate(user_id, input, existing)?;
 
-            if is_latest_doc(ctx, service_provider, &doc)
+            let document = service_provider
```
Comment (andreievg) [inline server/service/src/programs/encounter/update.rs:54]: does this relocation mean there is validation first before encounter row update ?

File: server/service/src/programs/patient/update.rs [hunk 8]
```diff
@@ -78,6 +73,15 @@ pub fn update_patient(
...
+
+                create_patient_name_store_join(&ctx.connection, store_id, &patient_id)?;
```
Comment (andreievg) [inline server/service/src/programs/patient/update.rs:82]: Just thinking sync thoughts:

It's much easier to reason about use case when create_patient_name_store_join only creates name_store_join when patient is created. I mean use cases of duplicates (does central server always sync name_store_join for patients ? or does it just sync the patients to reduce bandwidth ?).

File: server/service/src/sync/integrate_document.rs [hunk 9]
```diff
@@ -0,0 +1,87 @@
+use repository::{
```
Comment (andreievg) [inline server/service/src/sync/integrate_document.rs:1]: I am ignoring this in review, as I am not sure how it relates to document sync and doesn't seem to have any description of this in the PR.

File: server/service/src/sync/integrate_document.rs [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        .query_by_filter(
+            DocumentRegistryFilter::new().document_type(EqualFilter::equal_to(&document.r#type)),
+        )?
+        .pop()
+        .ok_or(RepositoryError::NotFound)?;
+
+    DocumentRepository::new(con).insert(document)?;
+
+    match registry.context {
+        repository::DocumentContext::Patient => update_patient(con, document)?,
```
Comment (andreievg) [inline server/service/src/sync/integrate_document.rs:29]: If there are multiple document for the same patient, does this mean we will be updating patient record each time ?

I thought your plan was to update the patient when document is opened ?

Does this PR account for circular sync ?

File: server/service/src/sync/integrate_document.rs
Comment (andreievg) [inline server/service/src/sync/integrate_document.rs:1]: I am not really sure what the difference is for document vs other record, I thought they would just sync in and the records would get upserted, or do we use them for syncing other records ? like program_enrolment ? 

Is there any documentation I can look up about document sync (that list all of the side effect of document syncing?)

File: server/service/src/sync/integrate_document.rs (hunk 10 above)
Comment (andreievg) [reply server/service/src/sync/integrate_document.rs:29]: Do we always sync documents in chronological order ?

File: server/service/src/sync/integrate_document.rs [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+        .map_err(|err| RepositoryError::as_db_error(&format!("{:?}", err), ""))?;
+    Ok(())
+}
+
+fn update_encounter(con: &StorageConnection, document: &Document) -> Result<(), RepositoryError> {
+    let (Some(patient_id), Some(program)) = (&document.owner_name_id, &document.context) else {
+        return Err(RepositoryError::as_db_error("Document owner id and context expected", ""));
+    };
+
+    let encounter = validate_encounter_schema(&document.data).map_err(|err| {
```
Comment (andreievg) [inline server/service/src/sync/integrate_document.rs:68]: in validate_encounter_schema, there is this comment:

https://github.com/openmsupply/open-msupply/blob/3649f446b69758dfd4fa39d88fa7ef33f14ceeb2/server/service/src/programs/encounter/validate_misc.rs#L27-L28

Encounter row seems to always have one defined of schema, I am quite confused does it mean we are using document as another layer, kind of like domain object but generic/dynamic domain object used to translate to typed data types, and this layer is again used in sync. So instead of syncing enrolments and encounters we sync documents and then translate them to enrolments and encounters ?

File: server/service/src/document/mod.rs (hunk 5 above)
Comment (clemens-msupply) [reply server/service/src/document/mod.rs:19]: to answer you comment better, it was working like this since every document version has a new id... However, there was the foot gun that this was assuming the document was already written which wasn't true in all cases.

File: server/service/src/programs/encounter/update.rs (hunk 7 above)
Comment (clemens-msupply) [reply server/service/src/programs/encounter/update.rs:54]: there was no reason for the re-position of the code. The reason simply that I didn't had the right data type when doing the sync where I had a `Document` and before this change it was working on with a `RawDocument`... So now I write the document and if it is the latest doc I also update the encounter table which seems also be the more nature flow.

File: server/service/src/sync/integrate_document.rs (hunk 10 above)
Comment (clemens-msupply) [reply server/service/src/sync/integrate_document.rs:29]: Yes the name table get updated every time when there is a new document version. It doesn't assumes chronological order for the incoming documents though.

> I thought your plan was to update the patient when document is opened ?

Yes its still the plan :)
https://github.com/openmsupply/open-msupply/issues/872

> Does this PR account for circular sync ?
is already handled in a later PR...

File: server/service/src/sync/integrate_document.rs (hunk 11 above)
Comment (clemens-msupply) [reply server/service/src/sync/integrate_document.rs:68]: The document is kind of the source of truth and yes for `patients`, `encounters`, and `program_enrolments` they must follow a predefined base schema.

This guaranteed base schema allows us to fill `encounter` and `program_enrolment` tables. These tables are basically just aux tables. For sync we only need to sync the documents and then fill the encounter and enrolment table from the received documents.

> Just had a brief thought, the dashboard on central, they would need to look at json data to figure out details about enrolment/encounter ?

yes this is correct, the queries would have to work on json data

File: server/service/src/sync/integrate_document.rs
Comment (clemens-msupply) [reply server/service/src/sync/integrate_document.rs:1]: There is no documentation for this yet. Does my other comment answer this question?

**Comment by clemens-msupply** [comment]
There where branch merges in the upper PRs which made review harder... I checked the upper PRs and they where relatively small, for this reason I merged them into this PR. Additional changes:

- Fixes document sync support (untested previously)
- Remove document comment column (not needed right now...)
- Update sync triggers (is_sync_update)
- Callback rename
- Fix event condition check

File: server/service/src/programs/patient/update.rs (hunk 8 above)
Comment (andreievg) [reply server/service/src/programs/patient/update.rs]: Yeah basically since the sync system is not easy to understand, and has quite a few rules and condition, and thus result in different states of data, I think it's a good idea to minimise possible use case. 

> If I understood correctly there could be the case where there we got a patient but not the matching name_store_join right?

I am actually not sure, I think they would sync, but if, in our code, we only create name_store_join when we create the name on remote site, then we have less use cases to consider

File: server/service/src/sync/integrate_document.rs (hunk 10 above)
Comment (andreievg) [reply server/service/src/sync/integrate_document.rs:31]: > Yes the name table get updated every time when there is a new document version. It doesn't assumes chronological order for the incoming documents though.

So during initialisation for example, we can't assume that name would be in the latest state. Since we get the latest state of name (as it itself syncs) why should we update_patient from document during sync ?

Comment (clemens-msupply) [reply server/service/src/sync/integrate_document.rs:31]: You mean before or after #872? Yes after #872 we don't need to update the name table here.

For now we probably do not need it either (apart from possible sync edge cases?)

File: server/service/src/sync/integrate_document.rs (hunk 11 above)
Comment (andreievg) [reply server/service/src/sync/integrate_document.rs:70]: > The document is kind of the source of truth and yes for patients, encounters, and program_enrolments they must follow a predefined base schema.

Ok, it makes sense, similar to the comment above, with confusion about syncing via document table, my concern was really about:

* With graphql we have a way to enforce schema (with inputs)
* With sync we have a way to validate and se/de to a concrete data type

Of course the above doesn't work for schema that is not know at runtime, but for schema that is know at runtime it has it's benefits. And I can see how documents can be used to extend a concrete data type, but what is the benefit of keeping the known schema (known at runtime), in the document, is it because it could be edited concurrently on two different sites ? Or because it's a generic way to store/sync it ?

File: server/graphql/programs/src/queries/patient.rs [hunk 12]
```diff
@@ -106,14 +106,9 @@ impl PatientNode {
...
 
-        let store_id = match self.patient.name_row.supplying_store_id.clone() {
-            Some(store_id) => store_id,
-            None => return Ok(None),
-        };
-
         let result = loader
             .load_one(DocumentLoaderInput {
-                store_id,
+                store_id: self.store_id.clone(),
```
Comment (andreievg) [inline server/graphql/programs/src/queries/patient.rs:111]: hmm saw this change and was looking a bit deeper into the loader, I can't find `store_id` used in the loader ¯\_(ツ)_/¯

File: server/repository/migrations/postgres/2022-03-27T10-00_update_changelog_upsert_with_sync_function/up.sql [hunk 13]
```diff
@@ -0,0 +1,9 @@
+CREATE FUNCTION update_changelog_upsert_with_sync()
```
Comment (andreievg) [inline server/repository/migrations/postgres/2022-03-27T10-00_update_changelog_upsert_with_sync_function/up.sql:1]: update_changelog_upsert_with_is_sync_update ?

That's pretty long thought, but `with_sync` is too ambiguous imo

File: server/repository/migrations/sqlite/2022-04-27T13-05_create_document_trigger/down.sql [hunk 14]
```diff
@@ -0,0 +1,3 @@
+DROP TRIGGER IF EXISTS document_insert_trigger;
+DROP TRIGGER IF EXISTS document_update_trigger;
+DROP TRIGGER IF EXISTS document_delete_trigger;
```
Comment (andreievg) [inline server/repository/migrations/sqlite/2022-04-27T13-05_create_document_trigger/down.sql:3]: Was wondering, when is document deletion allowed ? Can document be referenced by something that is not deleted, and when deleted causing errors somewhere (potentially of-site) -> runtime for constraint or logic if no constraints

There should really be a rule for only soft deleting central or `shared` data.

File: server/repository/migrations/postgres/2022-03-27T10-00_update_changelog_upsert_with_sync_function/up.sql (hunk 13 above)
Comment (andreievg) [reply server/repository/migrations/postgres/2022-03-27T10-00_update_changelog_upsert_with_sync_function/up.sql:1]: I think DB method works very nicely here, thanks

File: server/repository/migrations/sqlite/2022-04-27T13-05_create_document_trigger/down.sql (hunk 14 above)
Comment (andreievg) [reply server/repository/migrations/sqlite/2022-04-27T13-05_create_document_trigger/down.sql:3]: Also, can document be updated ?

File: server/service/src/document/mod.rs [hunk 15]
```diff
@@ -1,28 +1,24 @@
...
+/// Checks if the doc is the latest in the DB
 pub(crate) fn is_latest_doc(
     ctx: &ServiceContext,
     service_provider: &ServiceProvider,
-    doc: &RawDocument,
+    doc_name: &str,
+    doc_timestamp: DateTime<Utc>,
 ) -> Result<bool, RepositoryError> {
-    let latest_existing = service_provider
+    let Some(latest_existing) = service_provider
```
Comment (andreievg) [inline server/service/src/document/mod.rs:18]: Great the `let Some` works quite nice imo.

File: server/service/src/document/mod.rs [hunk 16]
```diff
@@ -1,28 +1,24 @@
...
-        if lastest_existing.timestamp > doc.timestamp {
-            // newer doc already exist
-            return Ok(false);
-        }
-    }
-    Ok(true)
+        .document(ctx, doc_name, None)? else {
+        return Ok(true);
+    };
+    Ok(latest_existing.datetime <= doc_timestamp)
```
Comment (andreievg) [inline server/service/src/document/mod.rs:23]: I think explicit sort would help (to guarantee that this wont' break if default sort is ever changed), would also help other validating the logic, I had to look up: https://github.com/openmsupply/open-msupply/blob/d8860428d38e45c382ab57a665c61801fdc7ece5/server/repository/src/db_diesel/document.rs#L271-L272

File: server/service/src/programs/encounter/encounter_updated.rs [hunk 17]
```diff
@@ -1,27 +1,23 @@
...
-
 use super::{
     encounter_schema::{self},
     validate_misc::ValidatedSchemaEncounter,
 };
 
 /// Callback called when the document has been updated
 pub(crate) fn update_encounter_row(
-    ctx: &ServiceContext,
+    con: &StorageConnection,
```
Comment (andreievg) [inline server/service/src/programs/encounter/encounter_updated.rs:14]: hmm, are there places where service_context not available where update_encounter_row is used ?

File: server/service/src/programs/encounter/insert.rs [hunk 18]
```diff
@@ -107,7 +82,38 @@ pub fn insert_encounter(
                     }
                 })?;
 
-            Ok(result)
+            if is_latest_doc(ctx, service_provider, &document.name, document.datetime)
```
Comment (andreievg) [inline server/service/src/programs/encounter/insert.rs:85]: Hmm, is there a way to `update` existing document ? I mean wouldn't is_latest_doc always be true ? Or is it for sync, or extra safety ?

File: server/service/src/programs/patient/patient_updated.rs [hunk 19]
```diff
@@ -11,12 +11,41 @@ use super::{
...
+            name_is_supplier: false,
+            is_sync_update: false,
+        })?;
+    }
+    Ok(())
+}
+
+/// Callback called when a patient document has been updated
+/// Updates the names table for the updated patient.
+pub fn update_patient_row(
```
Comment (andreievg) [inline server/service/src/programs/patient/patient_updated.rs:44]: I know it's more lines of code, but I really like the use of destructuring and spread in cases like this, due to exhaustive match and visibility of unused fields

```diff
diff --git a/server/service/src/programs/patient/patient_updated.rs b/server/service/src/programs/patient/patient_updated.rs
index 189710f74..95b41c892 100644
--- a/server/service/src/programs/patient/patient_updated.rs
+++ b/server/service/src/programs/patient/patient_updated.rs
@@ -44,31 +44,45 @@ pub fn create_patient_name_store_join(
 pub fn update_patient_row(
     con: &StorageConnection,
     update_timestamp: &DateTime<Utc>,
-    patient: SchemaPatient,
+    SchemaPatient {
+        code,
+        code_2,
+        contact_details,
+        date_of_birth,
+        first_name,
+        gender,
+        id,
+        is_deceased,
+        last_name,
+        middle_name: _,
+        notes: _,
+        passport_number: _,
+        socio_economics: _,
+        allergies: _,
+        birth_place: _,
+        health_center: _,
+        date_of_birth_is_estimated: _,
+        date_of_death: _,
+        family: _,
+    }: SchemaPatient,
     is_sync_update: bool,
 ) -> Result<(), UpdatePatientError> {
-    let contact = patient.contact_details.get(0);
-    let date_of_birth = match patient.date_of_birth {
+    let contact = contact_details.get(0);
+    let date_of_birth = match date_of_birth {
         Some(date_of_birth) => Some(NaiveDate::from_str(&date_of_birth).map_err(|err| {
             UpdatePatientError::InternalError(format!("Invalid date of birth format: {}", err))
         })?),
         None => None,
     };
-    let address = patient.contact_details.get(0);
+    let address = contact_details.get(0);
     let name_repo = NameRowRepository::new(con);
-    let existing_name = name_repo.find_one_by_id(&patient.id)?;
-    let existing_name = existing_name.as_ref();
+    let existing_name = name_repo.find_one_by_id(&id)?;
+
     name_repo.upsert_one(&NameRow {
-        id: patient.id.clone(),
-        name: patient_name(&patient.first_name, &patient.last_name),
-        code: patient.code.unwrap_or("".to_string()),
         r#type: NameType::Patient,
-        is_customer: existing_name.map(|n| n.is_customer).unwrap_or(true),
-        is_supplier: existing_name.map(|n| n.is_supplier).unwrap_or(false),
-        supplying_store_id: existing_name.and_then(|n| n.supplying_store_id.clone()),
-        first_name: patient.first_name,
-        last_name: patient.last_name,
-        gender: patient.gender.and_then(|g| match g {
+        name: patient_name(&first_name, &last_name),
+        is_deceased: is_deceased.unwrap_or(false),
+        gender: gender.and_then(|g| match g {
             SchemaGender::Female => Some(Gender::Female),
             SchemaGender::Male => Some(Gender::Male),
             SchemaGender::Transgender => Some(Gender::Transgender),
@@ -77,24 +91,28 @@ pub fn update_patient_row(
             SchemaGender::Unknown => Some(Gender::Unknown),
             SchemaGender::NonBinary => Some(Gender::NonBinary),
         }),
-        date_of_birth,
-        charge_code: existing_name.and_then(|n| n.charge_code.clone()),
-        comment: existing_name.and_then(|n| n.comment.clone()),
         country: address.and_then(|a| a.country.clone()),
         address1: address.and_then(|a| a.address_1.clone()),
         address2: address.and_then(|a| a.address_2.clone()),
         phone: contact.and_then(|c| c.phone.clone().or(c.mobile.clone())),
         email: contact.and_then(|c| c.email.clone()),
         website: contact.and_then(|c| c.website.clone()),
-        is_manufacturer: existing_name.map(|n| n.is_manufacturer).unwrap_or(false),
-        is_donor: existing_name.map(|n| n.is_donor).unwrap_or(false),
-        on_hold: existing_name.map(|n| n.on_hold).unwrap_or(false),
-        created_datetime: existing_name
-            .and_then(|n| n.created_datetime.clone())
-            .or(Some(update_timestamp.naive_utc())), // assume there is no earlier doc version
-        is_deceased: patient.is_deceased.unwrap_or(false),
-        national_health_number: patient.code_2,
+
+        id,
+        first_name,
+        last_name,
+        date_of_birth,
+        national_health_number: code_2,
         is_sync_update,
+        code: code.unwrap_or_default(),
+
+        ..existing_name.unwrap_or(NameRow {
+            is_customer: true,
+            is_deceased: false,
+            is_donor: true,
+            created_datetime: Some(update_timestamp.naive_utc()), // assume there is no earlier doc version
+            ..NameRow::default()
+        })
     })?;
 
     Ok(())
```

Above would suggest that the following fields are considered `common` omSupply fields that are not in mSupply, would they be common for all countries ? (or are we still to add patient schema to configurable schema and serialise this document to a NameRow type structure ?)

I still quite strong on keeping the know at runtime schema strongly typed (i.e. for both graphql and sync)

File: server/service/src/document/mod.rs (hunk 15 above)
Comment (andreievg) [reply server/service/src/document/mod.rs:18]: Although, just realised this looks like a simple `map`, sorry only suggesting this because it look like it was refactored due to stylistic preference
```rust
    Ok(service_provider
        .document_service
        .document(ctx, doc_name, None)?
        .map_or(false, |r| r.datetime <= doc_timestamp))
```

File: server/service/src/programs/update_program_document.rs [hunk 20]
```diff
@@ -173,19 +170,19 @@ fn is_truthy(value: &Value) -> bool {
     return true;
 }
 
-fn match_condition(condition: &EventCondition, doc: &RawDocument) -> bool {
+fn match_condition(condition: &EventCondition, doc: &Document) -> bool {
     let Some(field) = extract_value_field(&doc.data, &condition.field) else {
-        if condition.is_falsy.is_some() {
+        if condition.is_falsy.unwrap_or(false) {
```
Comment (andreievg) [inline server/service/src/programs/update_program_document.rs:175]: hmm, not really sure how this change relates to sync, but looks like it's a fix for something ? Just wondering could we catch this with tests ?

File: server/service/src/sync/integrate_document.rs [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+    is_sync_update: bool,
+) -> Result<(), RepositoryError> {
+    DocumentRepository::new(con).insert(document, is_sync_update)?;
+
+    let Some(registry) = DocumentRegistryRepository::new(con)
+        .query_by_filter(
+            DocumentRegistryFilter::new().document_type(EqualFilter::equal_to(&document.r#type)),
+        )?
+        .pop() else {
+        return Ok(());
```
Comment (andreievg) [inline server/service/src/sync/integrate_document.rs:27]: If we receive a document that is not in document registry, we quietly ignore it ? Shouldn't we have an error for this ? (it shoulnd't stop sync, but be captured in SyncBuffer)

File: server/service/src/sync/integrate_document.rs (hunk 10 above)
Comment (andreievg) [reply server/service/src/sync/integrate_document.rs:31]: Yeah would only be an issue if there are multiple documents (multiple edits), and we re-sync the site, a couple of other edge cases too, but shouldn't affect one site editing one name and not re-syncing

**Review by andreievg — APPROVED** [review]
Generally look good to me thanks, I think we should have more tests of documents updating related records correctly, I also had quite a few questions, main one about using documents for schema that is know at runtime, I think this would only be needed if we expect those documents to be concurrently edited on two site and as a result would require merge by field strategy

File: server/service/src/sync/integrate_document.rs (hunk 11 above)
Comment (clemens-msupply) [reply server/service/src/sync/integrate_document.rs:70]: yes the editing and generic sync is one advantage but you also have an easy way to store the history and you are also able to easily ensure document integrity (calculating the hash). Furthermore, you can easily add other document types without a known schema.

File: server/graphql/programs/src/queries/patient.rs (hunk 12 above)
Comment (clemens-msupply) [reply server/graphql/programs/src/queries/patient.rs]: thanks, yes its not needed anymore, I removed it

File: server/service/src/document/mod.rs (hunk 15 above)
Comment (clemens-msupply) [reply server/service/src/document/mod.rs:18]: yes would work as well (`.map_or(true,...)` though).

Site note, find `map_or` relative uncommon so couldn't understand the code without looking at the docs... had this as a general observation, functional style code is sometimes simply harder because you need to know the used lib much better...

File: server/service/src/programs/encounter/encounter_updated.rs (hunk 17 above)
Comment (clemens-msupply) [reply server/service/src/programs/encounter/encounter_updated.rs:14]: I changed it because it was used in sync and there was only a StorageConnection available (but probably no hard reason why it couldn't be a ServiceContext)

File: server/service/src/programs/encounter/insert.rs (hunk 18 above)
Comment (clemens-msupply) [reply server/service/src/programs/encounter/insert.rs:85]: There could be of some race condition where two users are editing a document at the same time, e.g. client 1 edits a document but the document only gets send to the server with a delay. In the meantime client 2 could submit a newer document.

Also in sync, documents could come out of order and only the latest document should update the encounter.

File: server/service/src/programs/patient/patient_updated.rs (hunk 19 above)
Comment (clemens-msupply) [reply server/service/src/programs/patient/patient_updated.rs:44]: Thanks yes agree with the destructuring. I updated the code.

The SchemaPatient is the common base schema which can still be adapted for different countries. E.g. this schema + country specific values such as birth order or handedness. Here, we receive the full patient document but we only extract the base schema for the integration.

Is there an option to have additional fields in GraphQL, e.g. a JSON input object with some predefined fields. This would be pretty neat for the API. You can do this in REST/swagger but haven't found something similar in GraphQL?
